### PR TITLE
Initialise batch norm stats for e2e models

### PR DESCRIPTION
**Pull request by AdamHillier**
## What do these changes do?

The default BatchNormalization moving-variance initialisation is `"ones"`, which makes sense for a full-precision convolution (because the weights are usually initialised with very small absolute value) but doesn't make sense for a binary convolution (where the absolute value of the weights is always 1). This causes instability in the end2end tests because it means that the magnitude of the convolution outputs is unrealistic. In places, we've attempted to correct for this by fiddling with the gamma/beta initialisation, but the real problem is the moving-variance statistics.

This PR adds ten steps of training with a low learning rate for each model in the e2e test (excluding QuickNet, which we load with trained weights). These few steps should result in more sensible values for the batch norm statistics, and regardless I think this makes the tests more relevent that using models with entirely random weights.

A nice benefit is that it appears that this allows us to reduce the absolute tolerance in the `np.allclose` comparison.

## How Has This Been Tested?

CI.

## Benchmark Results

N/A.

## Related issue number

N/A.

### CODE REVIEW

File: larq_compute_engine/tests/end2end_test.py [hunk 1]
```diff
@@ -152,35 +145,45 @@ def toy_model_int8(**kwargs):
...
 def preprocess(data):
-    return lqz.preprocess_input(data["image"])
+    return lqz.preprocess_input(data["image"]), data["label"]
 
 
 def assert_model_output(model_lce, inputs, outputs):
     interpreter = Interpreter(model_lce, num_threads=min(os.cpu_count(), 4))
     actual_outputs = interpreter.predict(inputs)
-    np.testing.assert_allclose(actual_outputs, outputs, rtol=0.001, atol=0.25)
+    np.testing.assert_allclose(actual_outputs, outputs, rtol=0.05, atol=0.125)
```
Comment (AdamHillier) [inline larq_compute_engine/tests/end2end_test.py:154]: I don't think it really made sense to have such a small rtol but huge atol (with rtol/atol 0.001/0.25, the rtol would only really be relevant for values with absolute value > 100), so I've bumped up the rtol as well as reducing the atol.

Comment (Tombana) [reply larq_compute_engine/tests/end2end_test.py:154]: I agree that `0.25` is huge. However it's what we observed for the keras-tflite difference for large models like Quicknet.
I think it might be better to pass a tolerance (or `Default`) to each model in `model_cls` so that we have normal tolerances everywhere and make it higher exclusively for QuickNet. What do you think?

Comment (AdamHillier) [reply larq_compute_engine/tests/end2end_test.py:154]: Yeah I agree, that's a good idea :+1:

File: larq_compute_engine/tests/end2end_test.py [hunk 2]
```diff
@@ -152,43 +131,67 @@ def toy_model_int8(**kwargs):
...
+        rtol = 0.05
+        atol = 0.2
+    elif model_cls == toy_model_int8:
+        # 0.025 atol is chosen because it allows an off-by-one error in the int8
+        # model (with +-3 scales) but not off-by-two.
+        rtol = 0.01
+        atol = 0.025
+    else:
+        rtol = 0.001
+        atol = 0.001
```
Comment (AdamHillier) [inline larq_compute_engine/tests/end2end_test.py:183]: @Tombana yeah you're right, turns out we can use *way* lower tolerances for the non-QuickNet models. And probably should do.

**Review by Tombana — APPROVED** [review]

File: larq_compute_engine/tests/end2end_test.py [hunk 3]
```diff
@@ -152,43 +130,67 @@ def toy_model_int8(**kwargs):
...
     [toy_model, toy_model_sequential, toy_model_int8, lqz.sota.QuickNet],
 )
 def test_simple_model(model_cls):
-    model = model_cls(weights="imagenet")
-    model_lce = convert_keras_model(
-        model, experimental_enable_bitpacked_activations=True
-    )
-
-    # Test on the flowers dataset
+    # Test on the OxfordFlowers dataset
```
Comment (lgeiger) [inline larq_compute_engine/tests/end2end_test.py:147]: This is not the same as oxford flowers, see: https://www.tensorflow.org/datasets/catalog/tf_flowers
```suggestion
    # Test on the TF flowers dataset
```

**Review by lgeiger — APPROVED** [review]
Nice!
